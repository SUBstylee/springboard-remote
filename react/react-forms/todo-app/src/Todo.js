import { useState } from "react";
import './Todo.scss';
const Todo = ({ task = 'default todo', id = '1', remove, update, completed, toggleCompleted }) => {
    const [edit, setEdit] = useState(task);
    const [isEdit, setIsEdit] = useState(false);

    const handleRemove = () => remove(id);
    const handleEdit = () => setIsEdit(isEdit ? false : true);
    const handleUpdate = (e) => {
        e.preventDefault();
        update(id, edit);
        setIsEdit(false);
    };
    const handleChange = (e) => setEdit(e.target.value);
    const handleToggle = (e) => toggleCompleted(id);

    let res;
    if (isEdit) {
        res = (
            <div className='Todo'>
                <form className="Todo-edit-form">
                    <input type='text' value={edit} onChange={handleChange} />
                    <button onClick={handleUpdate}>Save Changes</button>
                </form>
            </div>
        )
    } else {
        res = (
            <div className='Todo'>
                <li className={completed ? 'Todo-task completed' : 'Todo-task'} onClick={handleToggle}>{task}</li>
                <div className="Todo-buttons">
                    <button onClick={handleEdit}><i className="fas fa-pen" /></button>
                    <button onClick={handleRemove}><i className="fas fa-trash" /></button>
                </div>
            </div>
        );
    }
    return res;
};

export default Todo;
