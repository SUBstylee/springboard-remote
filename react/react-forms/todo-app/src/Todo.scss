.Todo{
    display: flex;
    margin: 0 -3rem 4px;
    padding: 1.1rem 3rem;
    justify-content: space-between;
    align-items: center;
    background: rgba($color: #fff, $alpha: .1)  
}
.Todo-edit-form{
    display: flex;
    flex-wrap: wrap;
    label{
        min-width: 100%;
        margin-bottom: .5rem;
        font-size: 1.3rem;
    };
    input{
        flex-grow: 1;
        border: none;
        background: #f7f1f1;
        padding: 0 1.5em;
        font-size: initial;
    }
    button{
        padding: 0 1.3rem;
        border: none;
        background: #f66;
        color: #fff;
        text-transform: uppercase;
        font-weight: bold;
        border: 1px solid rgba($color: #fff, $alpha: 0.3);
        margin-left: 5px;
        cursor: pointer;
        transition: background .2s ease-out;
    }
    button:hover{
        background: #b14747;
    }
    input, button{
        font-family: 'Helvetica', sans-serif;
        height: 3rem;
    }
    p{
        font-size: 0;
    }
}
.Todo-task{
    position: relative;
    transition: opacity 0.4s linear;
}
.completed{
    opacity: 0.6;
    color: #555;
    text-decoration: line-through rgba(207, 75, 75, 0.164);
}
.completed:before{
    content: '';
    position: absolute;
    top: 50%;
    left: -0.5rem;
    display: block;
    width: 0%;
    height: 2px;
    background: rgb(119, 3, 3);
    animation: strikeitem .6s ease-out 0s forwards;
}
.Todo-buttons{
    flex-shrink: 0;
    padding-left: .7em;
    button{
        border: none;
        font-size: 1em;
        margin: .4em;
        background: none;
        appearance: none;
        cursor: pointer;
        color: #fff;
    }
}; 
@keyframes strikeitem {
    to{
        width: calc(100% + 1rem);
    };
};