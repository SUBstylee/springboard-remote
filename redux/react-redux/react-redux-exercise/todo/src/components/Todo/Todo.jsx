import './Todo.scss';

import { useState } from 'react';

const Todo = ({ id, task, handleDelete, handleEdit }) => {
    const [edit, setEdit] = useState(task);
    const [isEditing, setIsEditing] = useState(false);

    const toggleEdit = () => setIsEditing(isEditing => !isEditing);

    const handleChange = (e) => setEdit(e.target.value);

    const updateChanges = (e) => {
        e.preventDefault();
        handleEdit(id, task);
        setIsEditing(false);
    };

    const deleteMeme = () => handleDelete(id);

    if (isEditing) {
        return (
            <div>
                <form onSubmit={updateChanges}>
                    <input
                        type='text'
                        value={edit}
                        onChange={handleChange}
                    />
                    <button>Stop Editing</button>
                </form>
            </div>
        );
    }

    return (
        <div>
            <button onClick={toggleEdit}>Edit</button>
            <button onClick={deleteMeme}>X</button>
            <p>{edit}</p>
        </div>
    );
};

export default Todo;