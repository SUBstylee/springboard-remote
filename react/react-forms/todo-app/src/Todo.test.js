import { render, fireEvent } from '@testing-library/react';
import Todo from './Todo';

it('renders without crashing', () => {
    render(<Todo />);
});

it('matches snapshot', () => {
    const { asFragment } = render(<Todo />);
    expect(asFragment()).toMatchSnapshot();
});

it('matches snapshot when editing', () => {
    const { asFragment, getByRole } = render(<Todo />);
    const editBtn = getByRole('button', { name: /edit/ });
    fireEvent.click(editBtn);
    expect(asFragment()).toMatchSnapshot();
});

it('updates on form submit', () => {
    const updateTodo = jest.fn();
    const { getByRole, getByText } = render(<Todo update={updateTodo} />);
    const editBtn = getByRole('button', { name: /edit/ });
    fireEvent.click(editBtn);
    const updateBtn = getByText('Save Changes');
    fireEvent.click(updateBtn);
    expect(updateTodo).toHaveBeenCalled();
});

it('deletes todo on delete button click', () => {
    const removeTodo = jest.fn();
    const { getByRole } = render(<Todo remove={removeTodo} />);
    const deleteBtn = getByRole('button', { name: /delete/ });
    fireEvent.click(deleteBtn);
    expect(removeTodo).toHaveBeenCalled();
});