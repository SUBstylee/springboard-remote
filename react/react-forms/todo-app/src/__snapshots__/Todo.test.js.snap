// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Todo"
  >
    <li
      class="Todo-task"
    >
      default todo
    </li>
    <div
      class="Todo-buttons"
    >
      <button
        aria-label="edit"
      >
        <i
          class="fas fa-pen"
        />
      </button>
      <button
        aria-label="delete"
      >
        <i
          class="fas fa-trash"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot when editing 1`] = `
<DocumentFragment>
  <div
    class="Todo"
  >
    <form
      class="Todo-edit-form"
    >
      <input
        type="text"
        value="default todo"
      />
      <button>
        Save Changes
      </button>
    </form>
  </div>
</DocumentFragment>
`;
